### PR TITLE
When terms in SProp do not unify, rely on proof irrelevance

### LINKED ISSUE #14370: Support for proof-irrelevance when unifying terms in SProp
#### Description of the problem

Unification does not seem to equate terms in `SProp` when those terms contain evars, as e.g. in:
```
Axiom A : SProp.
Axiom a : A.
Axiom P : A->Type.
Axiom f : A -> A.
Axiom g : A -> A.
Check fun x : P (f a) => x : P (g _).
(* The term "x" has type "P (f a)" while it is expected to have type "P (g ?a)". *)
```

Would it make sense, when unification has fails, to finally check if the terms weren't irrelevant and to succeed is they are (doing this e.g. in the `default` function of `evar_conv_x`)? Of course, the evars present in the terms would remain uninstantiated but it is still better than failing and we could also combine this with heuristics trying to automatically instantiate pending evars in `SProp` based on its type.

Note: see #14339 for experimenting relying on proof-irrelevance.

#### Coq Version

Versions supporting `SProp`, up to 8.13.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** experimental feature

Fixes #14370

- [ ] Added / updated test-suite
- [ ] Entry added in the changelog

### CODE REVIEW

File: pretyping/evarconv.ml [hunk 1]
```diff
@@ -557,9 +557,17 @@ let rec evar_conv_x flags env evd pbty term1 term2 =
...
         let term2 = apprec_nohdbeta flags env evd term2 in
         let default () =
+        match
           evar_eqappr_x flags env evd pbty
             (whd_nored_state env evd (term1,Stack.empty))
             (whd_nored_state env evd (term2,Stack.empty))
+        with
+        | UnifFailure _ as x ->
+           if Relevanceops.relevance_of_term env (EConstr.Unsafe.to_constr term1) == Sorts.Irrelevant ||
+              Relevanceops.relevance_of_term env (EConstr.Unsafe.to_constr term2) == Sorts.Irrelevant
```
Comment (SkySkimmer) [inline pretyping/evarconv.ml:567]: use the Retyping version instead

**Comment by herbelin** [comment]
Not sure exactly what should be extracted from bench (the latest table?). Strange that it seems to go slightly faster while `SProp` is rarely used.
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬──────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]   │   mem faults    │
│                             │                         │                                       │                                       │                          │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD   PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│                 coq-bignums │   59.63    60.19  -0.93 │   164883856562    166154463527  -0.76 │   209916291475    209913056564   0.00 │  463832   463900   -0.01 │   0    0    nan │
│  coq-performance-tests-lite │ 1674.02  1686.98  -0.77 │  4678708600459   4707742172907  -0.62 │  7446210177148   7452188998275  -0.08 │ 1904648  1904616    0.00 │   0    0    nan │
│       coq-engine-bench-lite │  337.10   339.04  -0.57 │   947649877080    951703361269  -0.43 │  1657315451606   1661971650383  -0.28 │ 3541064  3541196   -0.00 │   0    0    nan │
│                coq-bedrock2 │  332.19   333.82  -0.49 │   923491291449    926970183233  -0.38 │  1534019737644   1534073299465  -0.00 │  752804   752736    0.01 │   1    0    nan │
│              coq-verdi-raft │ 1241.68  1247.47  -0.46 │  3448386456704   3464629611954  -0.47 │  4615550973191   4614712943559   0.02 │  902336   902520   -0.02 │   0    0    nan │
│                    coq-corn │ 1572.46  1579.69  -0.46 │  4363633993688   4382096221307  -0.42 │  5986244679839   5986033446134   0.00 │  816948   817040   -0.01 │   0    0    nan │
│                 coq-coqutil │   62.26    62.51  -0.40 │   171665237219    172473446087  -0.47 │   214099978663    213661863864   0.21 │  597516   598252   -0.12 │   0    0    nan │
│                coq-rewriter │  725.60   728.07  -0.34 │  2017398464436   2024327210557  -0.34 │  2947521925931   2945645889326   0.06 │ 1101236  1101096    0.01 │   0    0    nan │
│            coq-math-classes │  187.81   188.36  -0.29 │   517778659129    518900502337  -0.22 │   652842042871    652843913832  -0.00 │  479732   479980   -0.05 │   0    0    nan │
│               coq-fourcolor │ 2710.71  2717.55  -0.25 │  7540918317223   7560336550278  -0.26 │ 12248703809002  12245924745217   0.02 │  730848   730640    0.03 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │  798.68   800.68  -0.25 │  2222990529698   2230021482486  -0.32 │  3309525741496   3309410761633   0.00 │ 1101084  1101056    0.00 │   0    0    nan │
│             coq-fiat-crypto │ 5117.83  5129.15  -0.22 │ 14238475653258  14266553862478  -0.20 │ 23932707337664  23918583288506   0.06 │ 2206120  2209764   -0.16 │   0    0    nan │
│                   coq-verdi │  106.02   106.22  -0.19 │   292412047117    292451661614  -0.01 │   387024097557    386251658303   0.20 │  540104   539384    0.13 │   0    0    nan │
│               coq-perennial │ 7255.70  7265.38  -0.13 │ 20173452528147  20201341453226  -0.14 │ 29078177339600  29086556855547  -0.03 │ 3128036  3113964    0.45 │   0    0    nan │
│                  coq-stdlib │  420.57   421.03  -0.11 │  1182903505643   1181124676610   0.15 │  1470218071635   1469982046383   0.02 │  566852   566700    0.03 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1133.40  1134.51  -0.10 │  3151157678852   3155311824582  -0.13 │  4712017854024   4707915807149   0.09 │  911896   911008    0.10 │   0    0    nan │
│                    coq-hott │  298.75   299.03  -0.09 │   811436983639    810381816183   0.13 │  1187616844446   1187487952608   0.01 │  583908   583912   -0.00 │   0    0    nan │
│                   coq-color │  526.28   526.77  -0.09 │  1460561713572   1463910475917  -0.23 │  1724699290586   1724032984844   0.04 │ 1125340  1105020    1.84 │   0    0    nan │
│      coq-mathcomp-ssreflect │   57.90    57.94  -0.07 │   160137553450    159571896354   0.35 │   189778045300    189734128289   0.02 │  522528   523072   -0.10 │   0    0    nan │
│            coq-fiat-parsers │  644.61   644.88  -0.04 │  1787975120347   1790971606413  -0.17 │  2555573835505   2555512792917   0.00 │ 3274472  3274500   -0.00 │   0    0    nan │
│          coq-mathcomp-field │  239.26   239.36  -0.04 │   665447064181    666290495918  -0.13 │   953661886132    953486585615   0.02 │  665996   663936    0.31 │   0    0    nan │
│      coq-mathcomp-character │  163.16   163.20  -0.02 │   453489242667    453860716775  -0.08 │   596538126065    596382438419   0.03 │  744564   744096    0.06 │   0    0    nan │
│               coq-fiat-core │  116.33   116.32   0.01 │   327209049463    327078663336   0.04 │   409796098833    408635075660   0.28 │  479688   479620    0.01 │   0    0    nan │
│       coq-mathcomp-solvable │  184.50   184.45   0.03 │   513393926648    513599668042  -0.04 │   680805504871    680655458968   0.02 │  683704   690040   -0.92 │   0    0    nan │
│                  coq-geocoq │ 1460.52  1460.12   0.03 │  4058536554287   4057504482022   0.03 │  5633897020631   5633178778675   0.01 │  992244   992296   -0.01 │   0    0    nan │
│        coq-mathcomp-algebra │  150.34   150.24   0.07 │   418279693547    417957254255   0.08 │   507666516258    507589840110   0.02 │  552740   552656    0.02 │   0    0    nan │
│             coq-lambda-rust │ 1361.46  1360.44   0.07 │  3785060295803   3787481734347  -0.06 │  5128472621387   5120121662136   0.16 │ 1110568  1122156   -1.03 │   0    0    nan │
│                 coq-unimath │ 4174.37  4168.73   0.14 │ 11607968880318  11599155388362   0.08 │ 21237903844588  21234558498017   0.02 │ 1005884  1121408  -10.30 │   0    0    nan │
│                    coq-core │  110.82   110.60   0.20 │   310158820658    310177564411  -0.01 │   456547048278    456542904143   0.00 │  295148   295160   -0.00 │   4    0    nan │
│              coq-coquelicot │   74.36    74.15   0.28 │   205243955325    204595762516   0.32 │   241337130792    240542483932   0.33 │  559896   559220    0.12 │   0    0    nan │
│                coq-coqprime │ 1491.30  1487.03   0.29 │  4145121988977   4139449474811   0.14 │  6590284197790   6589662530096   0.01 │ 1056296  1056416   -0.01 │   0    0    nan │
│       coq-mathcomp-fingroup │   52.29    52.06   0.44 │   144390475438    144634758398  -0.17 │   188865698639    188824835737   0.02 │  482192   482144    0.01 │   0    0    nan │
│                         coq │    4.54     4.50   0.89 │    12463427479     12492554417  -0.23 │    23420818394     23424320980  -0.01 │  276440   276436    0.00 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴──────────────────────────┴─────────────────┘
```

**Comment by ppedrot** [comment]
Per-line diff does not show anything suspicious either, which is a bit surprising given the patch.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Please add a test.

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Pull request merged**